reconcile/prometheus_rules_tester/integration.py

```diff
@@ -0,0 +1,278 @@
+import json
+import logging
+import sys
+from collections.abc import (
+    Iterable,
+    Mapping,
+)
+from dataclasses import dataclass
+from typing import (
+    Any,
+    Optional,
+)
+
+import yaml
+from sretoolbox.utils import threaded
+
+import reconcile.openshift_resources_base as orb
+from reconcile.gql_definitions.common.app_interface_vault_settings import (
+    AppInterfaceSettingsV1,
+)
+from reconcile.gql_definitions.common.prometheus_rules_resources import (
+    query as prr_query,
+)
+from reconcile.status import ExitCodes
+from reconcile.typed_queries.alerting_services_settings import get_alerting_services
+from reconcile.typed_queries.app_interface_vault_settings import (
+    get_app_interface_vault_settings,
+)
+from reconcile.utils import (
+    gql,
+    promtool,
+)
+from reconcile.utils.semver_helper import make_semver
+from reconcile.utils.structs import CommandExecutionResult
+
+# This comes from prometheus-operator. It is the largest configmap that they will
+# create. It is also the largest PrometheusRule yaml file they will process to add it
+# to the configmap.
+MAX_CONFIGMAP_SIZE = 0.5 * 1024 * 1024
+
+QONTRACT_INTEGRATION = "prometheus_rules_tester"
+QONTRACT_INTEGRATION_VERSION = make_semver(0, 1, 0)
+
+PROVIDERS = ["prometheus-rule"]
+
+NAMESPACE_NAME = "openshift-customer-monitoring"
+
+
+@dataclass
+class TestContent:
+    test_path: str
+    test: dict
+
+
+@dataclass
+class Test:
+    cluster: str
+    namespace_name: str
+    rule_path: str
+    rule: dict
+    rule_length: int
+    tests: Optional[list[TestContent]]
+    result: Optional[CommandExecutionResult] = None
+
+
+@dataclass
+class RuleToFetch:
+    namespace: dict[str, Any]
+    resource: dict[str, Any]
+
+
+# this is just for the early exit
+def get_prometheus_tests() -> dict[str, str]:
+    """Returns a path indexed dict with the prometheus tests content"""
+    gqlapi = gql.get_api()
+    query_func = gqlapi.query
+    tests = {}
+    for t in prr_query(query_func=query_func).tests:
+        tests[t.path] = t.content
+
+    return tests
+
+
+def fetch_rule_and_tests(
+    rule: RuleToFetch, vault_settings: AppInterfaceSettingsV1
+) -> Test:
+    """Fetches associated data from a ruke and builds a Test object"""
+    if "add_path_to_prom_rules" not in rule.resource:
+        rule.resource["add_path_to_prom_rules"] = False
+
+    openshift_resource = orb.fetch_openshift_resource(
+        resource=rule.resource, parent=rule.namespace, settings=vault_settings.dict()
```

Comment:
Safer to have `dict(by_alias=True)` so we are sure to convert to the query defined keys instead of the qenerate generated keys

Reply:
fixed in 370f1647